### ISSUE TITLE
configure DBGMCU on certain STM32 SOC

### ISSUE THREAD

**Issue by rlezuo1** [zephyrproject-rtos/zephyr#56096]
**Is your enhancement proposal related to a problem? Please describe.**

On the STM32H7 line of SOC debugging in Stop/Sleep mode needs DBGMCU to be configured, otherwise a debugger can not connect and must connect under reset (which destroys CPU state and is not desired).

**Describe the solution you'd like**
Add a KConfig option to configure DBGMCU unit to allow debugger attach. Proposed patch in comments.

```
diff --git a/soc/arm/st_stm32/common/soc_config.c b/soc/arm/st_stm32/common/soc_config.c
index 37cdfe7a5c..751f524eba 100644
--- a/soc/arm/st_stm32/common/soc_config.c
+++ b/soc/arm/st_stm32/common/soc_config.c
@@ -65,6 +65,8 @@ static int st_stm32_common_config(const struct device *dev)
 
 #endif /* CONFIG_USE_SEGGER_RTT */
 
+       HAL_EnableDBGSleepMode();       // works on STM32H7, needs Kconfig and what about other STM32 SOCs?
+
        return 0;
 }

```

**Comment by rlezuo1** [comment]
@erwango Yes, but the instance you pointed out is #ifdef'd on CONFIG_USE_SEGGER_RTT. 

I think what is needed is some CONFIG_SOC_SERIES_STM32XXX_DBGMCU_ENABLE (or similar) Kconfig option, or some more generic CONFIG_SOC_ALLOW_DEBUGGING which does the right thing(tm) on every SOC which has things like STM32's DBGMCU.

**Comment by erwango** [comment]
Ok, right, what about the following which is done on U5:
https://github.com/zephyrproject-rtos/zephyr/blob/b0e6a6f8a7257895cd9ac8033813a239aec717ef/soc/arm/st_stm32/stm32u5/power.c#L137-L139

**Comment by rlezuo1** [comment]
yes, this look absolutely right.

question is whether that something to be done for each and every STM32 variant or better goes into common, the SEGGER_RTT issues may even partially come from DBGMCU not being in a connectable state.

like this?
```
diff --git a/soc/arm/st_stm32/common/soc_config.c b/soc/arm/st_stm32/common/soc_config.c
index 37cdfe7a5c..ffe91b4cc3 100644
--- a/soc/arm/st_stm32/common/soc_config.c
+++ b/soc/arm/st_stm32/common/soc_config.c
@@ -57,13 +57,19 @@ static int st_stm32_common_config(const struct device *dev)
        LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_DBGMCU);
 #endif /* LL_APB1_GRP1_PERIPH_DBGMCU */
 
+#endif /* CONFIG_USE_SEGGER_RTT */
+
+#if defined(CONFIG_DEBUG) || defined(CONFIG_USE_SEGGER_RTT)
+       /* let STM32 SOC be attachable w/o reset in CONFIG_DEBUG
+         * SEGGER_RTT als depends on this
+         */
 #if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32MP1X)
        HAL_EnableDBGSleepMode();
 #else
        LL_DBGMCU_EnableDBGStopMode();
 #endif /* CONFIG_SOC_SERIES_STM32H7X || CONFIG_SOC_SERIES_STM32MP1X */
 
-#endif /* CONFIG_USE_SEGGER_RTT */
+#endif /* CONFIG_DEBUG || CONFIG_USE_SEGGER_RTT */
 
        return 0;
 }
```

**Comment by tagunil** [comment]
I'm also interested as I don't see much consistency in this matter between series and modes. I think we should cover DBGMCU configuration for Sleep/Stop/Standby modes for all the series somehow. It could be useful not just for debugging, but also for reflashing without hardware reset connected.

**Comment by rlezuo1** [comment]
@erwango: I'd create a pull-request, but I am not sure which route to go, there is power.c in place for: g0, g5, h7, l0, l4, l5, u5, wb and wl  (h7 is missing the CONFIG_DEBUG's DBGMCU configuration though). No power.c in f0, f1, f2, f3, f4, f7, g4, l1 and mp1.
So putting the CONFIG_DEBUG's DBGMCU into soc_common seems to be resonable in effort and effect. Bit plus: there already is the SEGGER_RTT fix in there, which is (hopefully) tested so the proposed patch in https://github.com/zephyrproject-rtos/zephyr/issues/56096#issuecomment-1479700684 should work. 
What do you think?

**Comment by erwango** [comment]
TBH, I'm not fully completely fine with this, because these are PM related files and up to now power related code is split in series. Until this is not the case anymore (@tagunil is working on that), I'd prefer not to put power related code in common.
Also your problem is related to PM on H7, which is not supported yet, so I prefer not modifying common series code for this.

This being said, I understand this part is missing some refinement and I agree it needs some care. So let's have a try and see where it goes.

Regarding the introduction of a new Kconfig symbol, please note that CONFIG_DEBUG is used today in stm32 pm related samples (and maybe tests ?) in order to be able to flash after PM tests with twister. This should  be taken into account if some changes are made in this area.

**Comment by rlezuo1** [comment]
@erwango @tagunil as this one came up again (on a stm32f7 this time) I've looked at the current implementation (v3.4.0), and things are basically still as they were back then.

Looking at the power.c implementation again, the enable debug mode is conditional on CONFIG_DEBUG for stm32's which have a power.c 

We recently had the issue to have to analyze a deadlock in a release build and of course attaching the debugger triggered a reset and the state was lost.

I really do think that CONFIG_DEBUG is wrong and something like CONFIG_DEBUG_STM32_ENABLE_DBGMCU_SLEEP is what you really want.

What do you think?

**Comment by tagunil** [comment]
> @erwango @tagunil as this one came up again (on a stm32f7 this time) I've looked at the current implementation (v3.4.0), and things are basically still as they were back then.
> 
> Looking at the power.c implementation again, the enable debug mode is conditional on CONFIG_DEBUG for stm32's which have a power.c 
> 
> We recently had the issue to have to analyze a deadlock in a release build and of course attaching the debugger triggered a reset and the state was lost.
> 
> I really do think that CONFIG_DEBUG is wrong and something like CONFIG_DEBUG_STM32_ENABLE_DBGMCU_SLEEP is what you really want.
> 
> What do you think?

Well, I don't think it should be blocked by my PM unification attempts, as this part doesn't really depend on the PM framework or the implementation. This problem should be solved once and for all, and I believe there is a way to have a common solution with a couple of ifdefs, but I could be wrong.
I think it's a good idea to have a separate config option for that, but the important point here is not to break existing use cases, where CONFIG_DEBUG is used to enable this functionality. Maybe a simple "select" will do.

**Comment by erwango** [comment]
Closing since https://github.com/zephyrproject-rtos/zephyr/pull/62647 has been merged
(@rlezuo1 , you can use `Fixes: #62647` to get issues closed automatically once PR are merged)
